File: template/packages/swift/Sources/{{Name}}Do/{{Name}}Client.swift
import Foundation
import RpcDo

/// Configuration options for {{Name}}Client
public struct {{Name}}ClientOptions: Sendable {
    /// API key for authentication
    public var apiKey: String?

    /// Base URL for the service (defaults to https://{{name}}.do)
    public var baseUrl: String

    /// Connection timeout
    public var timeout: TimeInterval

    public init(
        apiKey: String? = nil,
        baseUrl: String = "https://{{name}}.do",
        timeout: TimeInterval = 30.0
    ) {
        self.apiKey = apiKey
        self.baseUrl = baseUrl
        self.timeout = timeout
    }
}

/// {{Name}}.do client for interacting with the {{name}} service
///
/// Example:
/// ```swift
/// let client = {{Name}}Client(apiKey: ProcessInfo.processInfo.environment["DOTDO_KEY"])
/// try await client.connect()
/// defer { await client.disconnect() }
/// // Use the client...
/// ```
public actor {{Name}}Client {
    private let options: {{Name}}ClientOptions
    private var rpc: RpcClient?

    /// Creates a new {{Name}}Client with the given options
    public init(options: {{Name}}ClientOptions = {{Name}}ClientOptions()) {
        self.options = options
    }

    /// Creates a new {{Name}}Client with an API key
    public init(apiKey: String?, baseUrl: String = "https://{{name}}.do") {
        self.options = {{Name}}ClientOptions(apiKey: apiKey, baseUrl: baseUrl)
    }

    /// Whether the client is currently connected
    public var isConnected: Bool {
        rpc != nil
    }

    /// Connect to the {{name}}.do service
    @discardableResult
    public func connect() async throws -> RpcClient {
        if rpc == nil {
            var headers: [String: String] = [:]
            if let apiKey = options.apiKey {
                headers["Authorization"] = "Bearer \(apiKey)"
            }

            rpc = try await RpcClient.connect(
                to: options.baseUrl,
                headers: headers.isEmpty ? nil : headers
            )
        }
        return rpc!
    }

    /// Disconnect from the service
    public func disconnect() async {
        if let client = rpc {
            await client.close()
            rpc = nil
        }
    }

    /// Get the underlying RPC client (must be connected first)
    public func getRpc() throws -> RpcClient {
        guard let client = rpc else {
            throw {{Name}}Error.notConnected
        }
        return client
    }
}

File: template/packages/dotnet/src/{{Name}}Client.cs
using System.Net.WebSockets;
using Rpc.Do;

namespace {{Name}}.Do;

/// <summary>
/// Configuration options for {{Name}}Client
/// </summary>
public class {{Name}}ClientOptions
{
    /// <summary>
    /// API key for authentication
    /// </summary>
    public string? ApiKey { get; init; }

    /// <summary>
    /// Base URL for the service (defaults to https://{{name}}.do)
    /// </summary>
    public string BaseUrl { get; init; } = "https://{{name}}.do";

    /// <summary>
    /// Connection timeout
    /// </summary>
    public TimeSpan Timeout { get; init; } = TimeSpan.FromSeconds(30);
}

/// <summary>
/// {{Name}}.do client for interacting with the {{name}} service
/// </summary>
/// <example>
/// <code>
/// await using var client = new {{Name}}Client(new {{Name}}ClientOptions
/// {
///     ApiKey = Environment.GetEnvironmentVariable("DOTDO_KEY")
/// });
///
/// await client.ConnectAsync();
/// // Use the client...
/// </code>
/// </example>
public class {{Name}}Client : IAsyncDisposable
{
    private readonly {{Name}}ClientOptions _options;
    private RpcClient? _rpc;

    /// <summary>
    /// Creates a new {{Name}}Client with the given options
    /// </summary>
    public {{Name}}Client({{Name}}ClientOptions? options = null)
    {
        _options = options ?? new {{Name}}ClientOptions();
    }

    /// <summary>
    /// Creates a new {{Name}}Client with an API key
    /// </summary>
    public {{Name}}Client(string apiKey) : this(new {{Name}}ClientOptions { ApiKey = apiKey })
    {
    }

    /// <summary>
    /// Whether the client is currently connected
    /// </summary>
    public bool IsConnected => _rpc != null;

    /// <summary>
    /// Connect to the {{name}}.do service
    /// </summary>
    public async Task<RpcClient> ConnectAsync(CancellationToken cancellationToken = default)
    {
        if (_rpc == null)
        {
            var headers = new Dictionary<string, string>();
            if (!string.IsNullOrEmpty(_options.ApiKey))
            {
                headers["Authorization"] = $"Bearer {_options.ApiKey}";
            }

            _rpc = await RpcClient.ConnectAsync(
                _options.BaseUrl,
                headers: headers.Count > 0 ? headers : null,
                cancellationToken: cancellationToken
            );
        }
        return _rpc;
    }

    /// <summary>
    /// Disconnect from the service
    /// </summary>
    public async Task DisconnectAsync()
    {
        if (_rpc != null)
        {
            await _rpc.CloseAsync();
            _rpc = null;
        }
    }

    /// <summary>
    /// Get the underlying RPC client (must be connected first)
    /// </summary>
    public RpcClient Rpc => _rpc ?? throw new InvalidOperationException(
        "{{Name}}Client is not connected. Call ConnectAsync() first."
    );

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        await DisconnectAsync();
        GC.SuppressFinalize(this);
    }
}
